Implement ImGuiListClipper

jpwidera opened the pull request:
- Added dependency to Data.Vector
- Added Raw calls to ImGuiListClipper
- Added the withListClipper-Bracket notation (Albeit I'm currently a little bit uneducated about the varios LiftIO/MonadUnliftIOs)
- Added the test in the Main.hs

TristanCacqueray commented on dear-imgui.cabal:
```diff
@@ -236,11 +237,11 @@ executable test
   main-is: Main.hs
   default-language: Haskell2010
-  ghc-options: -Wall
+  ghc-options: -Wall 
```
nit: extra trailing space

jpwidera replied:
Ah, error from switching from and to debug mode. Btw: is there an alternative to manually edit the cabal file to enable
```
ghc-options: -Wall -g -rtsopts -dcore-lint -debug -O0
cc-options: -g -O0
```
or would it be better to implement a debug-mode?

dpwiz commented on src/DearImGui/Raw.hs:
```diff
@@ -245,6 +253,48 @@ C.context (Cpp.cppCtx <> C.bsCtx <> imguiContext)
 Cpp.using "namespace ImGui"
 
+-- | Wraps @ImGuiListClipper
+newtype ImGuiListClipper = ImGuiListClipper (Ptr ())
```
Add `ImGuiListClipper` to Structs.hs and wrap a pointer to it.

```haskell
newtype ListClipper = ListClipper (Ptr ImGuiListClipper)
```

This should help with `void*`'s and casts.

jpwidera replied:
Good idea!

dpwiz commented on src/DearImGui/Raw.hs:
```diff
@@ -245,6 +253,48 @@ C.context (Cpp.cppCtx <> C.bsCtx <> imguiContext)
+-- The pointer has to be deleted manually by destroiyImGuiListClipper
+beginListClipper :: MonadIO m => Int -> m ImGuiListClipper
+beginListClipper size 
+  | size < 0 = error "Size of the Clipper can't be negative"
```
Size would be better enforced by type system in non-raw wrapper by using `Natural` type.
Raw code should stick to C types, namely CInt and just pass it to the code.

dpwiz commented on src/DearImGui/Raw.hs:
```diff
@@ -245,6 +253,48 @@ C.context (Cpp.cppCtx <> C.bsCtx <> imguiContext)
+    let length = toEnum size :: CInt
+    ImGuiListClipper <$> [Cpp.block| void* { 
+        ImGuiListClipper* clipper = new ImGuiListClipper;
+        clipper->Begin( $(int length) );
```
I think `Begin` and `End` should be a methods of said clipper object.
Perhaps the clipper module can provide `new` and `delete` too.

Then `withListClipper` will assemble a whole routine from those basic blocks.

dpwiz replied:
*Everything* is unsafe in `Raw`.

dpwiz replied:
BTW, there's `IM_NEW` and `IM_DELETE` macros, presumably doing the right thing.

jpwidera replied:
Oho, ty for the info - didn't know, that ImGui actually allows custom allocators.

jpwidera replied:
Hm..it seems, that the IM_NEW is returning a null Ptr, if the allocating was not successfully and not throwing. I guess this should be checked somehow! Is there a standard way to check for NullPtr?

dpwiz replied:
I [did](https://github.com/haskell-game/dear-imgui.hs/blob/main/src/DearImGui.hs#L1693) it with `== Foreign.nullPtr`

jpwidera replied:
What do you think about throwing an error instead of adding an extra layer of Maybes. I mean, if allocation was not successfully there are other problems on the system. Eg:
```
  ptr <- [C.exp| ImGuiListClipper* { IM_NEW(ImGuiListClipper) } |]
  pure $ 
    if castPtr ptr == nullPtr
       then error "Could not allocate ImGuiListClipper"
       else ListClipper ptr
```

dpwiz replied:
I'm slightly against throwing exceptions and really wouldn't like a *bottom* handed to me.

Failed allocation is a bad sign indeed and there's not much one can do about it other than report and crash.
However, a context of said crash can be different from case to case.
Some would want a full-blown report, some would be okay using refutable pattern.

If anything, Maybe is an explicit signal that there's some shit you may have to deal with.

jpwidera replied:
Just two more cents:

1st. It makes the with-function a little bit more unreadable. It will add something like:
```
withListClipper :: (MonadUnliftIO m) => V.Vector a -> (a -> IO b) -> m ()
withListClipper list action = Raw.ListClipper.new >>= withListClipper' list action
  where
    withListClipper' _ _ Nothing = return ()
    withListClipper' list action (Just clipper) = liftIO do
      bracket (evaluate clipper) Raw.ListClipper.delete step
   ...
```
and 2nd: the bracket-function is still able to throw exceptions from the `action`.

Making a bullet-proof interface is very nice.On the other side, there would be much more extra-code to make each  c++-wrapper non-throwable. 

But if this is the way to go, I'll do :)

dpwiz replied:
I just found [throwIfNull](https://hackage.haskell.org/package/base-4.10.0.0/docs/Foreign-Marshal-Error.html#v:throwIfNull), making the wrapper straightforward:

```haskell
bracket new_ ListClipper.delete
  where
    new_ = do
      ListClipper ptr <- ListClipper.new -- unwrap raw pointer
      fmap ListClipper . liftIO $        -- wrap "certified not-null pointer"
        Foreign.throwIfNull "withListClipper: ListClipper allocation failed" ptr
```

It will be even simpler with newtype wrapper changed to an alias.
The pointers have all the distinction they need in their type and extra wrapping is _pointless_.

jpwidera replied:
>It will be even simpler with newtype wrapper changed to an alias.
>The pointers have all the distinction they need in their type and extra wrapping is pointless.

Looks good, if you are ok with colliding definition of pointers like:
```Haskell
newtype Font = Font (Ptr ImFont)
newtype Context = Context (Ptr ImGuiContext)
...
--vs.
type ListClipper = Ptr ImGuiListClipper
```

dpwiz replied:
We'll make a clean up round later.

ocharles commented on src/DearImGui.hs:
```diff
@@ -818,6 +824,23 @@ dragScalarN label dataType ref vSpeed refMin refMax format flags = liftIO do
+  
+-- | Clips a large list of items
+--
+-- The requirements on a are that they are all of the same height. Not sure, how this could be expressed.
```
```suggestion
-- The requirements on @a@ are that they are all of the same height.
```

jpwidera commented:
Ok, I think I addressed all issues from above.
I resolved all reviews that seemed trivial to me - is this your routine or should the reviewer resolve them?

dpwiz commented:
I'm preparing a pull PR to your PR with some minor stuff and docs...

dpwiz commented:
@jpwidera my cents :smile: 
https://github.com/jpwidera/dear-imgui.hs/pull/1
